services/rsources/handler.go

```diff
@@ -0,0 +1,154 @@
+package rsources
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lib/pq"
+)
+
+const (
+	tablePrefix = "job_run_stats_"
+)
+
+type sourcesHandler struct {
+	extension
+	jobRunIdTableMapLock sync.RWMutex
+	jobRunIdTableMap     map[string]struct{}
+}
+
+func (sh *sourcesHandler) GetStatus(ctx context.Context, jobRunId string, filter JobFilter) (JobStatus, error) {
+
+	statsTableName := fmt.Sprintf("%s%s", tablePrefix, jobRunId)
+	filterParams := []interface{}{}
+	filters := ``
+
+	if len(filter.TaskRunId) > 0 {
+		filters += ` task_run_id  = ANY ($1)`
+		filterParams = append(filterParams, pq.Array(filter.TaskRunId))
+	}
+
+	if len(filter.SourceId) > 0 {
+		if len(filterParams) > 0 {
+			filters += ` AND `
+		}
+		filters += fmt.Sprintf(` source_id = ANY ($%d)`, len(filterParams)+1)
+		filterParams = append(filterParams, pq.Array(filter.SourceId))
+	}
+
+	if len(filters) > 0 {
+		filters = ` WHERE ` + filters
+	}
+	sqlStatement := fmt.Sprintf(
+		`SELECT 
+			source_id,
+			destination_id,
+			task_run_id,
+			sum(in_count),
+			sum(out_count),
+			sum(failed_count) FROM "%[1]s" %[2]s
+			GROUP BY task_run_id, source_id, destination_id
+			ORDER BY task_run_id, source_id, destination_id DESC`,
+		statsTableName, filters)
+
+	rows, err := sh.extension.getReadDB().QueryContext(ctx, sqlStatement, filterParams...)
+
+	if err != nil {
+		var e *pq.Error
+		if err != nil && errors.As(err, &e) && e.Code == "42P01" {
+			return JobStatus{}, StatusNotFoundError
+		}
+		return JobStatus{}, err
+	}
+	defer rows.Close()
+
+	statMap := make(map[JobTargetKey]Stats)
+	for rows.Next() {
+		var statKey JobTargetKey
+		var stat Stats
+		err := rows.Scan(
+			&statKey.SourceId, &statKey.DestinationId, &statKey.TaskRunId,
+			&stat.In, &stat.Out, &stat.Failed,
+		)
+		if err != nil {
+			return JobStatus{}, err
+		}
+		statMap[statKey] = stat
+	}
+	if err := rows.Err(); err != nil {
+		return JobStatus{}, err
+	}
+
+	return statusFromQueryResult(jobRunId, statMap), nil
+}
+
+// IncrementStats checks for stats table and upserts the stats
+func (sh *sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, stats Stats) error {
+	err := sh.checkForTable(ctx, jobRunId)
+	if err != nil {
+		return err
+	}
+
+	jobRunIdStatTableName := fmt.Sprintf("%s%s", tablePrefix, jobRunId)
```

Comment:
This could lead to an SQL injection. Let's make sure we sanitise and validate `jobRunId` and test the attack surface. Using user-provided data as table names is tricky

Reply:
yes, as long as we make sure we remove any double quotes from the string, I don't see any other possibility for SQL injection.